lib/Hashing.hpp

```diff
@@ -0,0 +1,105 @@
+#ifndef PROTEUS_HASHING_HPP
+#define PROTEUS_HASHING_HPP
+
+#include "CompilerInterfaceTypes.h"
+#include "TimeTracing.hpp"
+
+#include <llvm/ADT/ArrayRef.h>
+#include <string>
+#if LLVM_VERSION_MAJOR == 18
+#include "llvm/ADT/StableHashing.h"
+#else
+#include "llvm/CodeGen/StableHashing.h"
+#endif
+
+namespace proteus {
+
+using namespace llvm;
+
+class HashT {
+private:
+  stable_hash Value;
+
+public:
+  inline HashT(const stable_hash HashValue) { Value = HashValue; }
+  inline HashT(const HashT &H) { Value = H.Value; }
+  inline HashT(StringRef S) { S.getAsInteger(0, Value); }
+  inline const stable_hash getValue() const { return Value; }
+  inline std::string toString() const { return std::to_string(Value); }
+  inline bool operator==(const HashT &Other) const {
+    return Value == Other.Value;
+  }
+};
+
+static HashT inline hashValue(HashT &H) { return H; }
+
+static HashT inline hashValue(StringRef &S) {
+  return stable_hash_combine_string(S);
+}
+
+static HashT inline hashValue(const std::string &S) {
+  return stable_hash_combine_string(S);
+}
+
+static HashT inline hashValue(unsigned long &V) {
+  return stable_hash_combine_string(
+      StringRef{reinterpret_cast<char *>(&V), sizeof(unsigned long)});
+}
+
+static HashT inline hashValue(unsigned int &V) {
+  return stable_hash_combine_string(
+      StringRef{reinterpret_cast<char *>(&V), sizeof(unsigned int)});
+}
```

Comment:
template <typename T>
```suggestion
static HashT inline hashValue(unsigned long &V) {
  return stable_hash_combine_string(
      StringRef{reinterpret_cast<char *>(&V), sizeof(unsigned long)});
}

static HashT inline hashValue(unsigned int &V) {
  return stable_hash_combine_string(
      StringRef{reinterpret_cast<char *>(&V), sizeof(unsigned int)});
}
```
typename std::enable_if<std::is_scalar<T>::value>::type
static HashT inline hashValue(T &V){
}